WordPress/openverse
# Add django-migration-git-conflict to prevent divergent migration merges

## Fixes



Fixes #687 by @sarayourfriend

## Description


Adds `django-migration-git-conflict` recommended [in this comment](https://github.com/WordPress/openverse/issues/687#issuecomment-2040480372) by @tylerlaprade (thanks!).

Check out the plugin's repository for information on how it works: https://github.com/crux-lab/django-migrations-git-conflicts?tab=readme-ov-file#how-it-works

To summarise: the latest migration files for each app will get updated when a new migration is added. When that file is merged, any other branches with divergent latest migrations will conflict, and need a rebase to fix.

Existing CI checks should all continue to work, and we can leave in the PR comment requesting a rebase for migrations anyway, as an additional helpful reminder. Ideally PRs with migrations are always rebased even without a conflict so that the migrated code is tested against main, in case there are issues with new code unaware of the migration changes (i.e., [a semantic conflict](https://martinfowler.com/bliki/SemanticConflict.html)). But we are always vulnerable to those, even outside the context of a migration, so I don't think we should consider them a special case with respect to the linked issue.

## Testing Instructions

Review the new files. Run `just api/dj makemigrations --check` locally and see that it works. Make a change to force a migration (change some help text on a model attribute, for example) and run makemigrations again, confirm that the latest migration is updated. All of that is really just showing how the plugin works, we don't need to test the plugin.

## Checklist


- [N/A] I added or updated tests for the changes I made (if applicable).
- [N/A] I added or updated documentation (if applicable).
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
